datastore/matcher_store.go

```diff
@@ -1,5 +1,11 @@
 package datastore
 
+// Iter is an iterator function that accepts a callback 'yield' to handle each
+// iterator item. The consumer can signal the iterator to break or retry by
+// returning an error. The iterator itself returns an error if the iteration
+// cannot continue or was interrupted unexpectedly.
+type Iter[T any] func(yield func(T) error) error
+
```

Comment:
I'd like to keep this aligned to the `rangefunc` experiment syntax (which looks likely to land in Go 1.23), so I'd rather we remove the error returns:
```suggestion
// Iter is an iterator function that accepts a callback 'yield' to handle each
// iterator item. The consumer can signal the iterator to break or retry by
// returning an error. The iterator itself returns an error if the iteration
// cannot continue or was interrupted unexpectedly.
type Iter[T any] func(yield func(T, error) bool)

```

Alternatively, if you feel strongly about always communicating the error, we could expand this a little:
```suggestion
type Iter[T any] func(yield func(T, error) bool)

type IntoIter[T any] func(/*ctx?*/) (Iter[T], func() error)
```
(This idea stolen from [here](https://github.com/golang/go/issues/65236#issuecomment-1953399166).)
And then the using code needs to call the second function (and the compiler should complain if that doesn't happen) and the implementer just needs to keep an `error` around to return for the second function to return.

Reply:
Cool, thanks for the context. Yielding errors suffice. I've updater the `Iter` interface. Here's an example of what users could do using the `jsonblob.Iter`:

```go
ref, err = u.store.UpdateVulnerabilitiesIter(ctx, it.Updater, it.Fingerprint, func(yield func(*claircore.Vulnerability, error) bool) {
	for it.Next() {
		if !yield(it.Vulnerability, nil) {
			return
		}
		count++
	}
	if loader.Err() != nil {
		_ = yield(nil, loader.Err())
	}
})
```